smithy-model/src/main/java/software/amazon/smithy/model/ShapeTypeFilteredSet.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model;
+
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import software.amazon.smithy.model.shapes.Shape;
+
+/**
+ * Provides a view of a set returned by the getXwithTrait methods so that only
+ * shapes of a specific type are treated as present within the set.
+ *
+ * @param <T> Type of shape to project out of the Set.
+ */
+final class ShapeTypeFilteredSet<T extends Shape> extends AbstractSet<T> {
+
+    private final Set<Shape> shapes;
+    private final Class<T> shapeType;
+    private int size;
+
+    ShapeTypeFilteredSet(Set<Shape> shapes, Class<T> shapeType) {
+        this.shapes = shapes;
+        this.shapeType = shapeType;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return o.getClass() == shapeType && super.contains(o);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new FilteredIterator<>(shapes.iterator(), shapeType);
+    }
+
+    @Override
+    public int size() {
+        // Compute and cache the number of elements in the set with the desired type.
+        int result = size;
+
+        // Note that a resulting size with no elements is inefficiently cached.
+        if (result == 0 && !shapes.isEmpty()) {
+            for (Shape shape : shapes) {
+                if (shapeType == shape.getClass()) {
+                    result++;
+                }
+            }
+            size = result;
+        }
+
+        return result;
+    }
+
+    private static final class FilteredIterator<T extends Shape> implements Iterator<T> {
+        private final Iterator<Shape> iterator;
+        private final Class<T> shapeType;
+        private T next;
+
+        FilteredIterator(Iterator<Shape> iterator, Class<T> shapeType) {
+            this.iterator = iterator;
+            this.shapeType = shapeType;
+
+            // Always compute the first element in the iterator.
+            next = computeNext();
+        }
+
+        @SuppressWarnings("unchecked")
+        private T computeNext() {
+            // Filter out shapes of other types.
+            while (iterator.hasNext()) {
+                Shape nextCandidate = iterator.next();
+                if (shapeType == nextCandidate.getClass()) {
+                    return (T) nextCandidate;
+                }
+            }
```

Comment:
What's the reason for skipping elements rather than creating a new set that only has those elements? Because it will likely only be run through once?

Reply:
This filtered view of the Set means that we don't have to create a new Set and preemptively copy all the values from one set to the other. Almost all the code that uses these methods (at least in our libraries) will just be iterating over the Set and not need the entire container populated up front.